Consolidate Docker configurations

jonseitz opened the pull request:
**I'm tagging everyone as a reviewer here because this PR introduces some breaking changes to the development environment.**

This PR does a fairly broad reconfiguration of the docker-related files (`Dockerfile`, `docker-compose.yml`, and `.env`). Though it touches a lot of different files, most of those changes are aimed at streamlining the configurations into single files wherever possible. Though there's no technical problem with having separate `.prod` and `.dev` files, this just makes it a little easier to get the project up and running from a fresh clone.

The changes to the Dockerfile are perhaps the most extensive. This new setup defined four separate stages that split into two separate branches that can be selected with the `--target` option -- a `development` image the has more dependencies installed on the base system and runs the code from the built-in Django webserver, and the `production` image that has only the minimal number of installed dependencies and runs the app with `gunicorn`. The goal here is to cut down on the total number of layers that need to be built, and to create a smaller final image (At last test the new `production` image weighs in at about 150MB, down from over 400MB previously).

I've also done a quick write through of a README that details the setup process and briefly describes the app architecture. There's still a gap to be filled re: the feed-import and indexing process, but I don't quite know that process well enough to fill it in. (@natalynyu Maybe you can write something?)

jonseitz committed:
Combine Dockerfiles

Instead of having separate .dev and .prod Dockerfiles, this creates a
single file with multiple targets. It also removes some of the system
dependencies from the final production image by copying all of the
installed/compiled dependencies from the builder stage, which brings the
final image size down to about 150MB (from 400MB previously).

You can build images from the individual targets by passing the
"--target" argument to "docker build", or by using the "target:" key
under "build:" in a docker-compose file.

Without any arguments, it will default to the final "production" image

jonseitz committed:
Consolidate entrypoint scripts

This creates a single entrypoint script where the final command changes
based on the presence or absence of a "--production" flag. This flag can
be passed in the CMD step of the Dockerfile as:

CMD ["--production"]

or in the "command" key of the "build:" section of a docker-compose file
with:

command: ["--production"]

This flag has been passed to the Dockerfile's production stage, so
omitting it entirely will default to the production mode

jonseitz committed:
Modify build script to use new Dockerfile

Changes the GitHub Actions workflow to use the top-level Dockerfile with
the ./app context, which will default to a production build.

jonseitz committed:
Rename dev compose file

As a minor quality of life improvement, this renames the
docker-compose.dev.yml to just docker-compose.yml so you no longer need
to pass the --file arugment to the cli. It also changes the build stage
to use our new unified Dockerfile with the appropriate target.

I've left the docker-compose.prod.yml here, updating the
"dockerfile:" and "command:" fields as necesary. As we'll be running the
app on ECS we won't need this for "actual" production, but it could be
useful to simulate a production environment in the future.

jonseitz committed:
Move Dockerfile into context

There were some symlink/path issues with having the Dockerfile outside
of the container build context, so I'm putting it back.

jonseitz committed:
Set ownership of media/static dirs

When I tried bringing up a fresh build of the production container I
started getting errors about missing permissions on the media and static
directories, which makes sense as the volumes are being mounted in the
nginx container as root-owned but in the django container as app-owned.
This adds an app user to the nginx container and gives that user owner
of media and static, so the permissions will match across both
containers.

jonseitz committed:
Use a single .env file

This is not a major change, but the .env file will be read automatically
by docker-compose without needing to be specified with the `env_file`
option, making it just that little bit easier to get started quickly.

I've moved the hard-coded differences between the two files (i.e DEBUG
and DJANGO_SETTINGS_MODULE`) directly into the docker-compose file so
that we don't need to maintain the separate templates, and removed the
POSTGRES_ specific environment variables in favor or defining them from
the SQL_ variables in the docker-compose. In development we need to have
the same user, password, and database defined in both, and this just
helps avoid any typos that could cause things to fail.

jonseitz committed:
Add Docker Documentation to the README

This provides a brief overview of the application as a whole and it's
connection to Algolia and to the sec-directory-client repo, as well as
more detailed instructions on how to get up and running initially.

Missing are the details about running the load_feed_people function and
indexing the data.

jonseitz committed:
Add _prod suffix to resources

This will tell docker-compose that the containers we use when running
the app in prod mode locally should not be the same as the dev
containers.

jonseitz committed:
Limit ports to localhost

This limits the ports opened by docker-compose to onl being reachable on
localhost, instead of being open to anyone on the same network. There's
minimal security risk with this application, but it's best practice to
always limit that access.

chrxr commented on docker-compose.prod.yml:
```diff
@@ -1,37 +1,44 @@
+      DEBUG: 0
+      DJANGO_SETTINGS_MODULE: app.settings.production
     depends_on:
       - db
```
Should this "depends_on" just be removed? It doesn't seem necessary, as it is isn't present in the Dev version, and it's currently breaking the build. Otherwise will need to change to `db_prod`.

jonseitz replied:
ah, excellent point. Dropped that line in 330eec0

chrxr commented on docker-compose.prod.yml:
```diff
@@ -1,37 +1,44 @@
       - db
-  db:
+
+  db_prod:
```
Should this just be db? Or is this so you could run a db and db_prod image at the same time? Currently it breaks the default environment variable for SQL_HOST.

jonseitz replied:
Yes, the idea was to have separate dev and prod containers, so I added an override of the `SQL_HOST` value in 1d67ccf.

chrxr commented on docker-compose.prod.yml:
```diff
@@ -1,37 +1,44 @@
-      - 1337:80
+      - 127.0.0.1:1337:80
     depends_on:
       - web
```
Either remove if not necessary or change to `web_prod`.

jonseitz replied:
Dropped in 330eec0

chrxr reviewed (changes requested):
Thanks @jonseitz. The only issues I found are around the naming of the docker-compose services in prod. As well as those mentioned in the comments, the nginx config references `web` and so should be updated to `web_prod`. Otherwise it looks great to me.

jonseitz committed:
Remove depends_on options:

These don't really seem to be necessary, and I had neglected to change
them to the prod-suffixed versions anyway.

jonseitz committed:
Override SQL_HOST in prod env

Since the name of the db container has been changed to db_prod, we need
to provide that name here in place of db.

jonseitz reviewed (commented):
Something weird I've noticed with the `docker-compose.prod.yml` setup is that bringing up http://localhost:1337 still shows the default splash screen and says that `DEBUG = True` is set in settings, even though as far as I can tell (i.e. with `python managepy diffsettings --all`) it is loading the correct `app.settings.production` file and `DEBUG` is being set to `False`. @chrxr Any idea what's happening there?

chrxr commented:
> Something weird I've noticed with the `docker-compose.prod.yml` setup is that bringing up http://localhost:1337 still shows the default splash screen and says that `DEBUG = True` is set in settings, even though as far as I can tell (i.e. with `python managepy diffsettings --all`) it is loading the correct `app.settings.production` file and `DEBUG` is being set to `False`. @chrxr Any idea what's happening there?

I'm not getting this. I'm getting an nginx 404 at the root http://localhost:1337/, as expected. Perhaps some kind of weird caching issue? If you type in an incorrect path, do you get the debug page, or a 404?

jonseitz committed:
Change to web_prod in nginx.conf

jonseitz commented:
Ah, OK. I completely dropped and rebuilt everything and it started working as expected. I think the problem may have been that I hadn't removed and rebuilt the `media` and `static` volumes, so maybe there was something cached in one of those.

I did find that the `nginx.conf` file needed another tweak (setting the upstream to `web_prod`) and pushed that in cbb8a8f.

natalynyu commented on README.md:
````diff
@@ -1 +1,53 @@
+```
+
+That will list all of the available commands, but for development the most important ones will be:
+
````
Perhaps we can add how to bring up Django Admin: 
```sh
# Accessing Django Admin:
$ docker-compose -f docker-compose.dev.yml up --build
# Go to http://localhost:8000/admin/
```

jonseitz replied:
Ah. I did have that in a paragraph under `### Development` but it probably should be broken out into a code block like the other commands. That's done in 00e91b9

natalynyu reviewed (changes requested):
Looks great! I just added in some suggestions for the README section.

jonseitz commented on README.md:
````diff
@@ -1 +1,53 @@
+
+```sh
+# Run tests
+$ docker-compose exec web python manage.py test
````
Ah right, so without this PR merged you'll have to pass the `--file docker-compose.dev.yml` argument, since the default is to use `docker-compose.yml`.

Both commands are doing the same thing, but I find the `docker-compose` version preferable because:

1. You don't have to pass `-it`, which I forget to do about half the time.
2. The name of the container is shorter and consistent (i.e. I have this repo cloned to subdirectory called `server` under `sec-directory`, so my container's called `server_web_1`)

natalynyu replied:
Ah I see, that makes sense, thanks!

jonseitz committed:
Move docker-compose up into a code block

The docker-compose up command was mentioned inline in this paragraph,
but given how the other commands in the README have been specified it
makes more sense to use a code block for this.

jonseitz committed:
Add load_feed_people command

This is the one-line method for running the load_feed_people function
from the python shell.

jonseitz committed:
Specify correct docker-compose file

jonseitz committed:
Add space after $ prompt

jonseitz commented on README.md:
````diff
@@ -1 +1,53 @@
+# Run database migrations
+$ docker-compose exec web python manage.py migrate
+```
+
````
Good idea. Add the one-line version of that in 0271a46

chrxr reviewed (approved):
Looks good to me. Thanks @jonseitz

natalynyu reviewed (approved):
Looks great! Approved.

jonseitz committed:
Add exec option

This changes from a boolean PRODUCTION value to an EXEC_MODE value that
can be set to production or development, but defaults to "exec". In exec
mode, it will run whatever is passed in as the command value so that we
can run the load_feed_people() function or other django-admin commands.
In the other modes, it will run the app in either production or
development mode as indicated, after running migrations and creating the
superuser.

jonseitz committed:
Add --development flag

This just sets the "--development" flag in the CMD directive, which
makes the development target default to running in development mode.

jonseitz commented:
Latest batch of commits modifies the `entrypoint.sh` script to allow for running arbitrary commands if anything other than `--production` or `--development` is given as the "command" argument. In practice, this will mimic the typical docker behavior, and allow us to run the `manage.py` commands as one-off tasks on ECS.

To test this, you can do the following:

```sh
# Build the docker image
$ docker build -t sec-directory-runner ./app/

# bring up the existing docker-compose file so we can use the db container
$ docker-compose up

# Execute manage.py -- the network command should point to the network created
# by your docker-compose file, as we'll still need to access the postgres container
$ docker run -it --rm --env-file .env --network sec-directory-server_default sec-directory-runner python manage.py
```

chrxr commented on README.md:
````diff
@@ -1 +1,62 @@
+$ docker-compose exec web python manage.py shell --command "from feedperson.utils import load_feed_people; load_feed_people()"
+```
+
+### Production
````
Should the readme be updated to include details of the new exec vs production vs development model? I.e. the commands that you put in the last comment re this pull request?

jonseitz replied:
Sure, that's added in 94ea487. I'm hoping that's one of those implementation details that won't too necessary for most use cases -- _it should just work_ -- but it's probably better to have that in there than not.

chrxr reviewed (changes requested):
This is working well for me. I think this is certainly a better way of working. Two comments:
1) Should the readme be updated to reflect the new situation re building the docker container for the different environments?
2) Is it worth noting somewhere that when using exec mode, the environment variables that were previously set in the docker-compose file will need to be explicitly added into the `.env` file. Specifically talking about db vs db_prod.

farassadek commented:
Can we add the following to the README
1. A simple note about the `Database` variable in the .env file.
2. how to execute the already built image with the env variables.

jonseitz committed:
Document template.env

Provides documentation for each of the environment variables that needs
to be set to run the app locally.

jonseitz committed:
Provide additional details on running in Docker

This expands the previous README to provide more information about the
different ways that the app can be run in Docker, including the usual
production/development modes and the "exec" mode for running arbitrary
commands (i.e., running the load_feed_people() function).

jonseitz committed:
Add note about DJANGO_SETTINGS_MODULE

This one's kind of important for running the app in production, so worth
calling out here in the README. Also adding it to our template.env with
a default to app.settings.dev, just to surface it.

jonseitz commented:
@farassadek I think I've addressed your notes in 9f23f73, 94ea487, and 38ca09e, but if there's anything else I overlooked please let me know.

farassadek reviewed (approved):
Looks good

chrxr reviewed (approved):
Looks good @jonseitz . Thanks!

natalynyu reviewed (approved):
LGTM - approved!
